readme
GNUSki 0.3 README
Copyright (C) 2007 Rudolf Olah
Licensed under the GNU GPL

== What is it? ==
A clone of Skifree which is a Windows game from the early to mid '90s.
It currently uses NCurses for graphics.

== What do you do? ==
Race downhill avoiding obstacles and the dreaded Abominable Snow Man.

== Controls ==

Left, H, h  - lean to the left
Right, L, l - lean to the right
Up, K, k    - slow down
Down, J, j  - speed up
Space       - trick (not available just yet)

== For Programmers ==

=== Libraries required ===
ncurses (compiled against ncurses-5.6 though any version should work fine)

=== Compiling ===
To compile GNUSki, run this command in the console:

  c++ -O2 -o gnuski main.c -lncurses

Or you can just type in `make` which will use the Makefile to compile GNUSki.


ChangeLog
GNUSki 0.3

2018-09-28  Marisa Giancarla  <marisag@synchronetbbs.org>
	* main.c: Writes out score after game is complete

2007-05-30  Rudolf Olah  <omouse@gmail.com>
	* object.c: Coloured trees green.	
	* main.c: Decreased refresh rate to counter-act the weird
	attron/attroff cursor crap.

2007-05-29  Rudolf Olah  <omouse@gmail.com>
	* main.c: Proper scrolling. Generates objects as they pass from
	view.

2007-05-28  Rudolf Olah  <omouse@gmail.com>
	* object.c: Facing is specified when drawing objects (there are a
	lot of objects that have the same facing so there's no need for
	the struct to store that info).
	* main.c: Added the proper controls and screen-scrolling.
	* main.c: Using a refresh rate now and a separate variable to
	control speed. Also keeping track of distance.	

2007-05-25  Rudolf Olah  <omouse@gmail.com>
	* main.c: Converted everything to C.
	* main.c: GNUSki is starting from version 0.1 all over again.

2006-04-19  Rudolf Olah  <omouse@gmail.com>
	* draw.h: Changed the commands around a bit. The clean-up has
	begun.


gnuski.pl
#!/usr/bin/perl
use strict;
use warnings;

use UI::Dialog;
use Term::ReadKey;
use Term::ANSIScreen qw(cls);
use Time::Piece;

my $StatsFileOutput = "/sbbs/doors/GNUSki-BBS/stats.txt";
my $ScoreFile = "/sbbs/doors/GNUSki-BBS/scores.txt";
my $GameEXE = "/sbbs/doors/GNUSki-BBS/gnuski";
my $GameScore = "/tmp/gnuski.score";
my $ClearCommand = "/usr/bin/clear";
my $SortCommand = "/usr/bin/sort";
my $MaxLines = 19;
my $HeadCommand = "/usr/bin/head -n $MaxLines";

###################################################
# No changes below here
###################################################

my $CR_ver = "1.0";

# Set UserName from command line
my $UserName = $ARGV[0];
if (!$UserName)
{
	print "You must supply the username as first argument!\n";
	exit 0;
}

my $d = new UI::Dialog ( backtitle => "GNUSki Version v$CR_ver", height => 20, width => 65, listheight => 5,
	order => [ 'ascii', 'cdialog', 'xdialog' ]);

my $windowtitle = "Welcome to GNUSki $UserName!";
my $enjoyedtitle = "We hope you enjoyed GNUSki $UserName!";

my $menuselection = "";

sub MainMenu
{
	$menuselection = $d->menu( title => 'Main Menu', text => 'Select one:',
                            list => [ '1', 'Play Game',
                                      '2', 'View Scores',
                                      'q', 'Quit' ] );
}

sub ProcessGame
{
	my $LastScore = 0;

	# Process the game scores
	# Pull in exit score
	open(my $fh, '<:encoding(UTF-8)', $GameScore) or die "Could not open file '$GameScore' $!";
	while (my $row = <$fh>)
	{
		chomp $row;
		$LastScore = $row;
	}
	close($fh);
	open(my $fh1, '>>', $StatsFileOutput) or die "Could not open file '$StatsFileOutput' $!";
	my $CurDate = localtime->strftime('%m/%d/%Y');
	print $fh1 "$LastScore\t$UserName\t$CurDate\n";
	close($fh1);
	system("$SortCommand -nrk1 $StatsFileOutput | $HeadCommand > lastscores");
	system("mv lastscores $StatsFileOutput");
	open(my $fh, '<', $StatsFileOutput) or die "Could not open file '$StatsFileOutput' $!";
	open(my $scoresfh, '>', $ScoreFile) or die "Could not open file '$ScoreFile' $!";
	while (my $row = <$fh>)
	{
		chomp $row;
		my ($Score, $Player, $Date) = split /\t/, $row;
		my $FormatString = sprintf("%5d %25s -  %s", $Score, $Player, $Date);
		print $scoresfh "$FormatString\n";
	}
	close($fh);
	close($scoresfh);
}

sub ViewScores
{
	if (! -f $ScoreFile)
	{
		system("$ClearCommand");
		print "No scores yet - play a game first!\n\n--- [ Press Return To Continue ] ---";
		my $input = <STDIN>;
		return;
	}
	open(my $fh, '<:encoding(UTF-8)', $ScoreFile) or die "Could not open file '$ScoreFile' $!";

	system("$ClearCommand");
	print "            GNUSki High Scores\n";
	print "---------------------------------------------\n";
	print "Score 			 Player    Date\n";
	print "---------------------------------------------\n";
 
	while (my $row = <$fh>)
	{
		chomp $row;
		print "$row\n";
	}
	close($fh);
	print "--- [ Press Return To Continue ] ---";
	my $input = <STDIN>;
}

while (-1)
{
	MainMenu();
	if (($menuselection eq "CANCEL") || ($menuselection eq "ESC") || ($menuselection eq "") || ($menuselection eq "q"))
	{
		$d->msgbox( title => $enjoyedtitle, text => "Thanks for playing..." );
		exit 0;
	}
	if ($menuselection eq "1")
	{
		# Run the game
		system("$GameEXE");
		ProcessGame();
	}
	elsif ($menuselection eq "2")
	{
		ViewScores();
	}
}

exit 0;


objects.c
/*  GNUSki - a clone of the old game Skifree where you race downhill
    avoiding rocks and trees and try to score points by doing some
    tricks.
    Copyright (C) 2007 Rudolf Olah

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "objects.h"

void
setupColors ()
{
  if (has_colors () == false)
    {
      endwin ();
      printf ("*** Your terminal does not support color ***\n");
      exit (1);
    }
  start_color ();
  init_pair (1, COLOR_GREEN, COLOR_BLACK);
  init_pair (2, COLOR_WHITE, COLOR_BLACK);
}

struct Object
makeObject (enum objectType type, int x, int y)
{
  struct Object o;
  o.type = type;
  o.x = x;
  o.y = y;
  return o;
}

void
setPosition (struct Object* o, int x, int y)
{
  o->x = x;
  o->y = y;
}

void
moveObject (struct Object* o, char facing, int speed)
{
  switch (facing)
    {
    case 'n': o->y -= speed; break;
    case 'e': o->x += speed; break;
    case 's': o->y += speed; break;
    case 'w': o->x -= speed; break;
    case '1': o->y -= speed; o->x -= speed; break;
    case '2': o->y -= speed; o->x += speed; break;
    case '3': o->y += speed; o->x += speed; break;
    case '4': o->y += speed; o->x -= speed; break;
    }
}

void
draw (struct Object o, char facing)
{
  switch (o.type)
    {
    case skier:
      switch (facing)
	{
	case 'n':
	case 's':
	  move (o.y, o.x);
	  printw ("||");
	  break;
	case 'w':
	  move (o.y, o.x);
	  printw ("==");
	  break;
	case 'e':
	  move (o.y, o.x+1);
	  printw ("==");
	  break;
	case '1':
	case '3':
	  move (o.y, o.x);
	  printw ("\\\\");
	  break;
	case '2':
	case '4':
	  move (o.y, o.x);
	  printw ("//");
	  break;
	}
      break;
    case tree:
      attron (COLOR_PAIR (1));
      move (o.y, o.x);
      printw ("/|\\");
      move (o.y+1, o.x);
      printw ("/|\\");
      move (o.y+2, o.x);
      printw (" | ");
      attroff (COLOR_PAIR (1));
      break;
    case rock:
      attron (COLOR_PAIR (2));
      move (o.y, o.x);
      printw ("o");
      attroff (COLOR_PAIR (2));
      break;
    case hill:
      move (o.y, o.x);
      printw ("/^\\");
    case none: default:
      break;
    }
  move (0, 0);
}

int
collision (struct Object player, struct Object target)
{
  switch (target.type)
    {
    case tree:
      return player.y == target.y+2 && player.x == target.x+1;
      break;
    case rock:
      return player.y == target.y && player.x == target.x;
      break;
    case none: default:
      return 0;
    }
  return 0;
}


objects.h
/*  GNUSki - a clone of the old game Skifree where you race downhill
    avoiding rocks and trees and try to score points by doing some
    tricks.
    Copyright (C) 2007 Rudolf Olah

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef OBJECTS_H
#define OBJECTS_H

#include <stdio.h>
#include <stdlib.h>
#include <ncurses.h>

enum objectType { none, tree, rock, hill, skier, bigfoot };

struct Object
{
  enum objectType type;
  int x, y;
};

struct Object makeObject (enum objectType type, int x, int y);
void setPosition (struct Object* o, int x, int y);

/* Move the object in a certain facing/direction */
/* facing is one of the following characters:
   1 n 2
   w   e
   4 s 3
*/
void moveObject (struct Object* o, char facing, int speed);

void draw (struct Object o, char facing);

/* Checks for collision between a player object and a target */
int collision (struct Object player, struct Object target);

#endif


main.c
/*  GNUSki - a clone of the old game Skifree where you race downhill
    avoiding rocks and trees and try to score points by doing some
    tricks.
    Copyright (C) 2007 Rudolf Olah

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "objects.h"
#include <stdlib.h>
#include <time.h>
#include <stdio.h>
#include <ncurses.h>

#define MAX_OBJECTS 300
#define REFRESH_RATE 5000

int main (int argc, char* argv[])
{
  struct Object player, objects[MAX_OBJECTS];
  unsigned int c = 0, i = 0, maxRows, maxCols, ticker = 0, score = 0,
    distance = 0, speed = 1, style = 0;
  enum mode {loop, lose, win} state = loop;
  char facing[2] = {'s', 'n'};   /* Player facing, object facing */

  srand ((unsigned)time (NULL)); /* Seed random-number generator */
  initscr ();             /* Start ncurses */
  cbreak ();              /* Allow control breaks (Ctrl+C) */
  noecho ();              /* Don't echo characters to screen */
  keypad (stdscr, true);  /* Allow cursor key usage */
  curs_set (0);           /* Hide the cursor */
  getmaxyx (stdscr, maxRows, maxCols); /* Get coordinates of the terminal */
  setupColors ();

  /* Create objects */
  player = makeObject (skier, maxCols/2, maxRows/2);
  for (i = 0; i < MAX_OBJECTS; i++)
    objects[i] = makeObject (rand () % 3+1,
			     rand () % maxCols,
			     rand () % (maxRows*4 + maxRows/2));

  /* Menu will go here */
  printw ("GNUSki 0.3 - Skifree clone using NCurses, licensed under the GNU GPL.\n");
  printw ("Press any key to start...");
  refresh ();
  getch ();
  nodelay (stdscr, true); /* No waiting for input from the user */
  while (state == loop)
    {
      switch (c)
	{
	case KEY_LEFT: case 'h': case 'H':
	  facing[0] = '4';
	  facing[1] = '2';
	  break;

	case KEY_RIGHT: case 'l': case 'L':
	  facing[0] = '3';
	  facing[1] = '1';
	  break;

	case KEY_UP: case 'k': case 'K':
	  if (speed > 0)
	    speed--;
	  break;

	case KEY_DOWN: case 'j': case 'J':
	  if (speed < 3)
	    speed++;
	  break;

	case 'Q': case 'q':
	  state = lose;
	  break;
	}

      if (ticker == REFRESH_RATE)
	{
	  clear ();
	  ticker = 0;

	  /* Check for collisions and draw the objects */
	  for (i = 0; i < MAX_OBJECTS; i++)
	    {
	      if (objects[i].type != none && collision (player, objects[i]))
		state = lose;
	      moveObject (&objects[i], facing[1], speed);
	      if (objects[i].y < 0)
		setPosition (&objects[i], rand () % (maxCols*2),
			     rand () % maxRows + maxRows);
	      if (objects[i].x > 2 && objects[i].x < maxCols-2
		  && objects[i].y > 2 && objects[i].y < maxRows-2)
		draw (objects[i], facing[1]);
	    }
	  draw (player, facing[0]);
	  printw ("Time:   0:00:00.00\n");
	  printw ("Dist:     %02im\n", distance);
	  printw ("Speed:    %02im/s\n", speed);
	  printw ("Style:    %4i", style);
	  distance += speed;
	}
      refresh ();
      if (speed > 0)
	ticker++;
      c = getch ();
    }
  clear ();
  if (state == win)
    printw ("You won.");
  else
    printw ("You lost.");
  printw ("Your score was %i. You traveled %im.\nPress any key to continue...",
	  distance+style, distance);
  
FILE *f = fopen("/tmp/gnuski.score", "w");
if (f == NULL)
{
    printf("Error opening score file!\n");
    exit(1);
}
fprintf(f, "%d\n", distance+style);
fclose(f);
refresh ();
  nodelay (stdscr, false);
  getch ();

  endwin ();
  return 0;
}
